ports/cheriot-rtos/machine_uart.c

```diff
@@ -0,0 +1,205 @@
+
+    uint32_t baudrate = args[ARG_baudrate].u_int;
+    uart_init(self->uart_block, baudrate);
+}
+
+
+static mp_obj_t mp_machine_uart_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, MP_OBJ_FUN_ARGS_MAX, true);
+
+    int id = mp_obj_get_int(args[0]);
+    volatile open_titan_uart_t *block = get_uart_block(id);
+
+    machine_uart_obj_t *self = m_new_obj(machine_uart_obj_t);
+    self->base.type = &machine_uart_type;
+    self->uart_block = block;
+
+    mp_map_t kw_args;
+    mp_map_init_fixed_table(&kw_args, n_kw, args + n_args);
+    mp_machine_uart_init_helper(self, n_args - 1, args + 1, &kw_args);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+
+// TODO this one too
+static void mp_machine_uart_deinit(machine_uart_obj_t *self) {
+
+}
+
+static mp_int_t mp_machine_uart_any(machine_uart_obj_t *self) {
+    return uart_get_rx_level(self->uart_block);
+}
+
+static bool mp_machine_uart_txdone(machine_uart_obj_t *self) {
+    return uart_get_tx_level(self->uart_block) == 0;
+}
+
+static mp_uint_t mp_machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t size, int *errcode) {
+    machine_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    uint8_t *dest = (uint8_t *)buf_in;
+
+    for (size_t i = 0; i < size; i++) {
+        uint32_t timeout = (i == 0? self->timeout_ms:self->char_timeout_ms);
+        if (!uart_read(self->uart_block, &dest[i], timeout)) {
+            for (size_t x = 0; x < size; x++) {
+            }
+            if (i <= 0) {
+                *errcode = MP_EAGAIN;
+                return MP_STREAM_ERROR;
+            } else {
+                return i;
+            }
+        }
+
+    }
+    return size;
+}
+
+
+static mp_uint_t mp_machine_uart_write(mp_obj_t self_in, const void *buf_in, mp_uint_t size, int *errcode) {
+    machine_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    uint8_t *data = (uint8_t *)buf_in;
+
+    for (size_t i = 0; i < size; i++) {
+        uart_write(self->uart_block, data[i]);
+    }
+
+    return size;
+}
+
+
+static mp_uint_t mp_machine_uart_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_t arg, int *errcode) {
+    machine_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_uint_t ret;
+
+    switch (request) {
+        case MP_STREAM_POLL: {
+            uintptr_t flags = arg;
+            ret = 0;
+            if ((flags & MP_STREAM_POLL_RD) && uart_is_readable(self->uart_block)) {
+                ret |= MP_STREAM_POLL_RD;
+            }
+            if ((flags & MP_STREAM_POLL_WR) && uart_is_writable(self->uart_block)) {
+                ret |= MP_STREAM_POLL_WR;
+            }
+            break;
+        }
+        // case MP_STREAM_FLUSH:
+        //     break;
+
+        default: {
+            *errcode = MP_EINVAL;
+            ret = MP_STREAM_ERROR;
+            break;
+        }
+    }
+    return ret;
+}
+
+
+#define MICROPY_PY_MACHINE_UART_CLASS_CONSTANTS
```

Comment:
`#define` at the bottom of the file -- does this actually do anything?

Reply:
seems not - I've removed it